### PR TITLE
DDF-4168 Adds Size enforcer plugin to fail builds for artifacts that grow too large

### PR DESCRIPTION
**Pull request by coyotesqrl**
#### What does this PR do?
Adds size enforcer plugin defined in codice/ddf-support#54.

Many of our artifacts are already well over the default size threshold of
1Mb; for the purposes of this PR, the thresholds for those modules have been set
higher.

#### Who is reviewing it? 

@peterhuffer @kcwire @bakejeyner 

#### Select relevant component teams: 
N/A

#### Ask 2 committers to review/merge the PR and tag them here.

@lessarderic
@tbatie
@vinamartin

#### How should this be tested?

- CI will ensure that nothing is broken
- Pulling down the branch and setting the threshold for an artifact lower than its actual size should fail the build 

#### What are the relevant tickets?
There is no Jira ticket for this. This work is being performed against a GitHub Issue.
Fixes: #3774 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/core/catalog-core-camelcomponent/pom.xml [hunk 1]
```diff
@@ -170,6 +170,35 @@
                     </instructions>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-enforcer-plugin</artifactId>
+                <version>1.4.1</version>
+                <dependencies>
```
Comment (paouelle) [inline catalog/core/catalog-core-camelcomponent/pom.xml]: ✏️ Move this dependency part in the parent pom plugins management section where you can define this plugin with its version and dependencies. This way, you won't have to repeat it in every pom files.

Comment (coyotesqrl) [reply catalog/core/catalog-core-camelcomponent/pom.xml]: I was going to say there isn't much advantage to doing that, but it is eight lines that can be removed from any module's pom that needs to specify its own threshold. Good call.

**Comment by coyotesqrl** [comment]
As for your comment about the parent setting a default...I could, but the plugin itself has a built-in default of 1Kb which I think is good. That covers ~350 modules that don't have the plugin section added to them at all.

These ~80 modules are all above (or right at the cusp) of 1Kb.

I had considered using `combine.self="override"` for this, but I don't think it would save much or anything at all in this particular case.

**Comment by tbatie** [comment]
It'd be great to capture a couple metrics before this is merged. Is there an easy way to see how many modules are being skipped because the artifact to check isn't being resolved? It would be good to filter that list by modules that have a packing of jar to ensure the plugin is running on everything as expected.

I'm also interested in how fast this plugin is running. Can we record how long it takes to run this across the repo? I doubt this is impacting the build time much but you can never be too safe.

**Comment by coyotesqrl** [comment]
I ran three quick builds locally 
1. With size enforcer: 10:02
2. Without size enforcer: 11:11
3. With enforcer again: 10:30

I ran the third build with the thought that Maven may have pulled some artifacts down that were needed for the build in the first of the sequence above, inflating its duration. Looking at these numbers, I'd say that the use of the artifact enforcer is minimal enough to fall within the variation of build lengths.

**Comment by coyotesqrl** [comment]
build now

**Comment by troymohl** [comment]
What is the intended usage of this check? If a PR goes over the configured maximums, does that module need to be refactored before accepting the change?

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
@bdeining We're experimenting with GitHub Issues and duplicating the issue between the two tools is busy work.

**Comment by bdeining** [comment]
It also seems like busy work to manage issues in two separate areas.

**Comment by coyotesqrl** [comment]
Boldly migrating all of our issues over from Jira without first vetting the suitability of the platform would be rash.

**Comment by bdeining** [comment]
I agree; that's why i'm surprised we wouldn't stick to our primary mechanism for tracking issues while we investigate using a different one.

**Comment by bdeining** [comment]
curious what the answer to @troymohl 's question is as well

**Comment by coyotesqrl** [comment]
@troymohl Sorry. Missed your comment earlier.

The intent of this is similar to Jacoco coverage numbers. We can certainly raise the allowed size of a generated artifact, just as we can lower coverage thresholds, but there should be justification for doing so other than expediency. It's a good way for developers and reviewers to get a clear indication that someone has newly embedded some single huge jar or embedded a large number without good cause.

The size of our distribution is directly and negatively impacted by the growing cumulative size of all our artifacts and can best be addressed by minimizing and extracting embedded dependencies and relying more on OSGi's design.

**Comment by coyotesqrl** [comment]
Duplicated the GH Issue as a Jira ticket.

File: pom.xml [hunk 2]
```diff
@@ -56,7 +56,7 @@
         <snapshots.repository.url/>
         <reports.repository.url/>
         <!-- Must MANUALLY update this if the ddf/support project's version changes -->
-        <ddf.support.version>2.3.14</ddf.support.version>
+        <ddf.support.version>2.3.15-SNAPSHOT</ddf.support.version>
```
Comment (lessarderic) [inline pom.xml]: ❓Should we release `ddf-support` instead of having a dependency on a `SNAPSHOT` version?

Comment (coyotesqrl) [reply pom.xml]: Yes. This won't release until then. It's just referencing the snapshot until the release of support is completed.

**Comment by coyotesqrl** [comment]
@lessarderic I definitely meant 1Mb in my comment. I've updated the PR description and commit message.

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
build now
These are build-time changes, verifying only the build artifact. If packaging of the individual artifacts succeeds, it is correct. Any itest (or unit test) failures are unrelated.

**Comment by coyotesqrl** [comment]
build now
unrelated failsafe error

**Pull request merged**